Library/Homebrew/test/livecheck/strategy/sparkle_spec.rb

```diff
@@ -7,91 +7,137 @@
+        <enclosure os="osx" url="#{item_hashes[:v121][:url]}" sparkle:shortVersionString="#{item_hashes[:v121][:short_version]}" sparkle:version="#{item_hashes[:v121][:version]}" length="12345678" type="application/octet-stream" sparkle:dsaSignature="ABCDEF+GHIJKLMNOPQRSTUVWXYZab/cdefghijklmnopqrst/uvwxyz1234567==" />
       </item>
     EOS
 
-    appcast = <<~EOS
-      <?xml version="1.0" encoding="utf-8"?>
-      <rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:sparkle="http://www.andymatuschak.org/xml-namespaces/sparkle">
-        <channel>
-          <title>Example Changelog</title>
-          <link>#{appcast_url}</link>
-          <description>Most recent changes with links to updates.</description>
-          <language>en</language>
-          #{first_item}
-          #{second_item}
-        </channel>
-      </rss>
+    v120_item_with_macos_os = <<~EOS
+      <item>
+        <title>#{item_hashes[:v120][:title]}</title>
+        <sparkle:minimumSystemVersion>10.10</sparkle:minimumSystemVersion>
+        <sparkle:releaseNotesLink>https://www.example.com/example/#{item_hashes[:v120][:short_version]}.html</sparkle:releaseNotesLink>
+        <pubDate>#{item_hashes[:v120][:pub_date]}</pubDate>
+        <enclosure os="macos" url="#{item_hashes[:v120][:url]}" sparkle:shortVersionString="#{item_hashes[:v120][:short_version]}" sparkle:version="#{item_hashes[:v120][:version]}" length="12345678" type="application/octet-stream" sparkle:dsaSignature="ABCDEF+GHIJKLMNOPQRSTUVWXYZab/cdefghijklmnopqrst/uvwxyz1234567==" />
+      </item>
     EOS
 
-    omitted_items = appcast.sub("</item>", "</item>\n#{items_to_omit}")
-    beta_channel_item = appcast.sub(
-      first_item,
-      first_item.sub(
-        "</title",
-        "</title>\n<sparkle:channel>beta</sparkle:channel>",
-      ),
+    # This main `appcast` data is intended as a relatively normal example.
+    # As such, it also serves as a base for some other test data.
+    appcast = create_appcast_xml <<~EOS
+      #{v123_item}
+      #{v122_item}
+      #{v121_item_with_osx_os}
+      #{v120_item_with_macos_os}
+    EOS
+
+    omitted_items = create_appcast_xml <<~EOS
+      #{v123_item.sub(%r{<(enclosure[^>]+?)\s*?/>}, '<\1 os="not-osx-or-macos" />')}
+      #{v123_item.sub(/(<sparkle:minimumSystemVersion>)[^<]+?</m, '\1100<')}
+      <item>
+      </item>
+    EOS
+
+    beta_channel_item_string = v123_item.sub(
+      /<sparkle/, "<sparkle:channel>beta</sparkle:channel><sparkle"
     )
-    no_versions_item =
-      appcast
-        .sub(second_item, "")
-        .gsub(/sparkle:(shortVersionString|version)="[^"]+?"\s*/, "")
-        .sub(
-          "<title>#{item_hash[0][:title]}</title>",
-          "<title>Version</title>",
-        )
-    no_items = appcast.sub(%r{<item>.+</item>}m, "")
-    undefined_namespace = appcast.sub(/\s*xmlns:sparkle="[^"]+?"/, "")
+
+    beta_channel_item = appcast.sub(v123_item, beta_channel_item_string)
```

Comment:
```suggestion
    beta_channel_item = appcast.sub(
      v123_item,
      v123_item.sub(
        "</title>",
        "</title>\n<sparkle:channel>beta</sparkle:channel>",
      ),
    )
```

I'm still in favor of inlining the value of `beta_channel_item_string`, since we don't reuse it (i.e., I don't think it's worth creating a variable to simplify the appearance of the code). If we don't like the nested multiline `#sub` calls, it's technically possible to have the `v123_item.sub` call on one line while still passing `brew style` (i.e., our line limit is 120 characters and this was just to make things a bit more readable for people like me who wrap at 80 characters).

Besides that, I slightly prefer anchoring the substitution to the closing `</title>` tag (as in the existing substitution), as that's common throughout all of the item XML examples in these tests.

Reply:
This is one of the substitutions that for whatever reason I just find hard to follow without the named intermediary. It's difficult for me to follow the double layer of substitutions and I don't think the named intermediary adds much noise for the sake of clarity.

That said I don't mind using the closing `</title>` as the anchor.

Reply:
The way I think of it is "set `beta_channel_item` to a copy of `appcast` where the first item has `<sparkle:channel>beta</sparkle:channel>` appended after `</title>`". Any difficulty in reading it may arise from using a multiline `v123_item.sub` call within the multiline `appcast.sub` call, so I wouldn't be against the following alternative formatting if it helps to make it more understandable for you:

```suggestion
    # Set the first item in a copy of `appcast` to the "beta" channel, to test
    # filtering items by channel using a `strategy` block.
    beta_channel_item = appcast.sub(
      v123_item,
      v123_item.sub("</title>", "</title>\n<sparkle:channel>beta</sparkle:channel>"),
    )
```

With the added comment in this suggestion, I don't see this as something that's likely to trip up most readers. I still don't see the `beta_channel_item_string` change as potentially beneficial enough to warrant its inclusion, so I will have to pass on it. Hopefully this suggestion is more amenable to you but, either way, one of these two suggestions will need to be incorporated before this is merged.

Reply:
It's a shame we couldn't come to terms but I appreciate the effort you put in to reviewing things. I guess in the interest of getting the logical change in I'll go one of the suggestions and drop the discussion. :)